crates/gas-price-algorithm/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+}
+
+impl AlgorithmV1 {
+    pub fn new(
+        da_max_change_percent: u8,
+        min_da_price: u64,
+        da_p_value_factor: i64,
+        da_d_value_factor: i64,
+        moving_average_window: i64,
+        exec_change_amount: u64,
+        min_exec_price: u64,
+    ) -> Self {
+        Self {
+            da_max_change_percent,
+            min_da_price,
+            da_p_value_factor,
+            da_d_value_factor,
+            moving_average_profit: RefCell::new(0),
+            last_profit: RefCell::new(0),
+            profit_slope: RefCell::new(0),
+            moving_average_window,
+            exec_change_amount,
+            min_exec_price,
+        }
+    }
+
+    pub fn calculate_da_gas_price(
+        &self,
+        old_da_gas_price: u64,
+        total_production_reward: u64,
+        total_da_recording_cost: u64,
+    ) -> u64 {
+        let new_profit = total_production_reward as i64 - total_da_recording_cost as i64;
+        self.calculate_new_moving_average(new_profit);
+        self.calculate_profit_slope(*self.moving_average_profit.borrow());
+        let avg_profit = *self.moving_average_profit.borrow();
+        let slope = *self.profit_slope.borrow();
+
+        let max_change = (old_da_gas_price
+            .saturating_mul(self.da_max_change_percent as u64)
+            / 100) as i64;
+
+        // if p > 0 and dp/db > 0, decrease
+        // if p > 0 and dp/db < 0, hold/moderate
+        // if p < 0 and dp/db < 0, increase
+        // if p < 0 and dp/db > 0, hold/moderate
+        let p_comp = avg_profit / self.da_p_value_factor;
+        let d_comp = slope / self.da_d_value_factor;
+        let pd_change = p_comp + d_comp;
+        let change = min(max_change, pd_change.abs());
+        let sign = pd_change.signum();
+        let change = change * sign;
+        let new_da_gas_price = old_da_gas_price as i64 - change;
+        max(new_da_gas_price, self.min_da_price as i64) as u64
+    }
+
+    pub fn calculate_exec_gas_price(
+        &self,
+        old_exec_gas_price: u64,
+        used: u64,
+        capacity: u64,
+    ) -> u64 {
+        // TODO: This could be more sophisticated, e.g. have target fullness, rather than hardcoding 50%.
+        let new = match used.cmp(&(capacity / 2)) {
+            std::cmp::Ordering::Greater => {
+                old_exec_gas_price.saturating_add(self.exec_change_amount)
+            }
+            std::cmp::Ordering::Less => {
+                old_exec_gas_price.saturating_sub(self.exec_change_amount)
+            }
+            std::cmp::Ordering::Equal => old_exec_gas_price,
+        };
+        max(new, self.min_exec_price)
+    }
+
+    fn calculate_new_moving_average(&self, new_profit: i64) {
+        let old = *self.moving_average_profit.borrow();
+        *self.moving_average_profit.borrow_mut() = (old
+            * (self.moving_average_window - 1))
+            .saturating_add(new_profit)
+            .saturating_div(self.moving_average_window);
```

Comment:
Hmm, it seems that after restarting the node `self.moving_average_profit` and `self.last_profit` will be zero, which means the algorithm will work differently. We need to persist all information required for the algorithm.

Hmm, it would be nice if we used one iteration formula to avoid a persistent state. But I assume you haven't found any suitable for us=)

Reply:
Or maybe 

```rust
        old_da_gas_price: u64,
        old_total_production_reward: u64,
        old_total_da_recording_cost: u64,
```

can be stored inside of the algorithm and passed during creation. Then we should be able to calculate profit form them.

Reply:
This is a great point. Configuring of that might be a pain if the services aren't running when we start the algorithm, but it would be a solution. But how do we know what values it died on?

Another solution is we have an assoc type for the algo trait that the algo returns with arbitrary metadata. The type needs to match some data storage on the provider at compilation time, and we can have that type be versioned so the data storage can still be backwards compatible. Seems like a lot of trouble lol.

```
impl<FB, DA, A, GP> FuelGasPriceProvider<FB, DA, A, GP>
where
    FB: FuelBlockHistory,
    DA: DARecordingCostHistory,
    A: GasPriceAlgorithm<Metadata=<Self as GasPriceHistory>::Metadata>,
    GP: GasPriceHistory,
{
```
ooooor we can add storage to the algorithm--which is just so many traits 
:turtle: 
:turtle: 
:turtle: 
:turtle: 

I'd suggest that we just merge the algo abstraction into the provider, but I think that would make versioning harder. 

I feel like there is a cleaner solution here.